language/complexity-check-plugin/src/complexity-check.ts

```diff
@@ -0,0 +1,328 @@
+import {
+  Diagnostic,
+  DiagnosticSeverity,
+  Range,
+} from "vscode-languageserver-types";
+import { TextDocument } from "vscode-languageserver-textdocument";
+import { resolveDataRefs } from "@player-ui/player";
+import {
+  getProperty,
+  isPropertyNode,
+  AssetASTNode,
+  toRange,
+} from "@player-tools/json-language-service";
+import type {
+  ASTNode,
+  ASTVisitor,
+  PlayerLanguageService,
+  PlayerLanguageServicePlugin,
+  ViewASTNode,
+  DocumentContext,
+} from "@player-tools/json-language-service";
+
+const makeRange = (
+  start: number,
+  end: number,
+  document: TextDocument
+): Range => {
+  return {
+    start: document.positionAt(start),
+    end: document.positionAt(end),
+  };
+};
+
+export interface ComplexityCheckConfig {
+  /** Cutoff for content to be acceptable */
+  maxAcceptableComplexity: number;
+  /**
+   * If set, any score above this number but below `maxAcceptableComplexity` will be logged as a warning
+   */
+  maxWarningLevel?: number;
+  /** If set, maps complexity based on asset or view type */
+  typeWeights?: Record<string, number>;
+}
+
+/**
+ * Estimates complexity of content
+ */
+export class ComplexityCheck implements PlayerLanguageServicePlugin {
+  name = "complexity-check";
+
+  private config: ComplexityCheckConfig;
+  private typeCount: Record<string, number>;
+  private contentScore: number;
+  private range: {
+    start: number;
+    end: number;
+  };
+  private verboseDetails: Diagnostic[];
+
+  constructor(config: ComplexityCheckConfig) {
+    this.config = config;
+    this.typeCount = {};
+    this.contentScore = 0;
+    this.range = { start: 0, end: 0 };
+    this.verboseDetails = [];
+  }
+
+  apply(service: PlayerLanguageService): void {
+    service.hooks.validate.tap(this.name, async (_ctx, validation) => {
+      validation.useASTVisitor(this.createContentChecker(_ctx));
+    });
+
+    service.hooks.onDocumentUpdate.tap(this.name, () => {
+      this.typeCount = {};
+      this.contentScore = 0;
+      this.verboseDetails = [];
+    });
+
+    service.hooks.onValidateEnd.tap(this.name, (diagnostics, context) => {
+      const diagnosticRange = {
+        start: { line: 0, character: 0 },
+        end: { line: 0, character: 0 },
+      };
+
+      // TO DO: Potentially display this in a totals summary within diag-renderer
```

Comment:
Can we pull this dead code into a TODO ticket or a working branch for this feature? I would rather not have commented out code merged into the main branch